[RLlib; docs] RLlib documentation do-over (new API stack): Main index page. (#48285)
ray-project/ray: doc/source/rllib/rllib-algorithms.rst

    :width: 750

>     **PPO architecture:** In a training iteration, PPO performs three major steps: sampling a set of episodes or episode fragments (1),

Reviewer: ```suggestion
    **PPO architecture:** In a training iteration, PPO performs three major steps: 
    1. Sampling a set of episodes or episode fragments
```

>     converting these into a train batch and updating the model using a clipped objective and multiple SGD passes over this batch (2),

Reviewer: ```suggestion
    1. Converting these into a train batch and updating the model using a clipped objective and multiple SGD passes over this batch
```

>     and synching the weights from the Learners back to the EnvRunners (3).

Reviewer: ```suggestion
    1. Synching the weights from the Learners back to the EnvRunners
```

.. figure:: images/algos/dqn-architecture.svg
    :width: 650

>     **DQN architecture:** DQN uses a replay buffer to temporarily store episode samples collected from the environment.

Reviewer: ```suggestion
    **DQN architecture:** DQN uses a replay buffer to temporarily store episode samples that RLlib collects from the environment.
```

.. figure:: images/algos/sac-architecture.svg
    :width: 750

>     **SAC architecture:** SAC uses a replay buffer to temporarily store episode samples collected from the environment.

Reviewer: ```suggestion
    **SAC architecture:** SAC uses a replay buffer to temporarily store episode samples that RLlib collects from the environment.
```

.. figure:: images/algos/impala-architecture.svg
    :width: 750

    **IMPALA architecture:** In a training iteration, IMPALA requests samples from all EnvRunners asynchronously and the collected episodes
    are returned to the main algorithm process as Ray references rather than actual objects available on the local process.
    IMPALA then passes these episode references to the Learners for asynchronous updates of the model.
    RLlib doesn't always synch back the weights to the EnvRunners right after a new model version is available.

>     To account for the resulting off-policy'ness on the EnvRunners, IMPALA uses a procedure called v-trace,

Reviewer: ```suggestion
    To account for the EnvRunners being off-policy, IMPALA uses a procedure called v-trace,
```

    :width: 750

>     **IMPALA architecture:** In a training iteration, IMPALA requests samples from all EnvRunners asynchronously and the collected episode

Reviewer: ```suggestion
    **IMPALA architecture:** In a training iteration, IMPALA requests samples from all EnvRunners asynchronously and RLlib returns the collected episode
```

>     samples are returned to the main algorithm process as ray references (rather than actual objects available on the local process).

Reviewer: If you could convert these to active voice to clarify who is doing the returning or passing. Is it RLlib?

>     samples are returned to the main algorithm process as ray references (rather than actual objects available on the local process).

Reviewer: ```suggestion
    samples to the main algorithm process as Ray references rather than actual objects available on the local process.
```

>     These episode references are then passed to the Learners for asynchronous updates of the model.

Reviewer: ```suggestion
    RLlib then passes these episode references to the Learners for asynchronous updates of the model.
```

    To account for the fact that this asynchronous design leads to some degree of off-policiness

>     on the EnvRunners (models aren't always synched back to EnvRunners right after a new version of the weights is available), IMPALA uses

Reviewer: ```suggestion
    on the EnvRunners. RLlib doesn't always sync back the models to EnvRunners immediately after a new version of the weights is available. IMPALA uses
```

    **APPO architecture:** APPO is an asynchronous variant of :ref:`Proximal Policy Optimization (PPO) <ppo>` based on the IMPALA architecture,
    but using a surrogate policy loss with clipping, allowing for multiple SGD passes per collected train batch.

>     In a training iteration, APPO requests samples from all EnvRunners asynchronously and the collected episode

Reviewer: ```suggestion
    In a training iteration, APPO requests samples from all EnvRunners asynchronously and RLlib returns the collected episode
```

Author: I feel like it's super obvious that the `EnvRunners` return the requested samples, b/c of the "APPO requests samples from **all EnvRunners** ... ".

Mentioning that "RLlib" does the returning is way too general. Everything is run by RLlib, so this doesn't give the user any useful information as to which particular component does the returning.

Reviewer: "RLlib" was my guess and the point was just to clarify who's doing the returning, if needed. If it's obvious to the reader, just ignore my suggestion.

>     samples are returned to the main algorithm process as ray references (rather than actual objects available on the local algo process).

Reviewer: ```suggestion
    samples to the main algorithm process as Ray references rather than actual objects available on the local algorithm process.
```

>     These episode references are then passed to the Learner for asynchronous updates of the model.

Reviewer: ```suggestion
    RLlib then passes these episode references to the Learner for asynchronous updates of the model.
```

    To account for the fact that this asynchronous design leads to some degree of off-policiness

>     on the EnvRunners (models aren't always synched back to EnvRunners right after a new version of the weights is available), APPO uses

Reviewer: ```suggestion
    on the EnvRunners. RLlib doesn't always sync models back to EnvRunners immediately after a new version of the weights is available. APPO uses
```

    **APPO architecture:** APPO is an asynchronous variant of :ref:`Proximal Policy Optimization (PPO) <ppo>` based on the IMPALA architecture,
    but using a surrogate policy loss with clipping, allowing for multiple SGD passes per collected train batch.
    In a training iteration, APPO requests samples from all EnvRunners asynchronously and the collected episode
    samples are returned to the main algorithm process as Ray references rather than actual objects available on the local process.
    APPO then passes these episode references to the Learners for asynchronous updates of the model.
    RLlib doesn't always synch back the weights to the EnvRunners right after a new model version is available.

>     To account for the resulting off-policy'ness on the EnvRunners, APPO uses a procedure called v-trace,

Reviewer: ```suggestion
    To account for the EnvRunners being off-policy, APPO uses a procedure called v-trace,
```

Reviewer: Just trying to get around saying "off-policy'ness". Please correct as needed.

Author: Ok, makes sense. I'll review ...

    :width: 850

    **DreamerV3 architecture:** DreamerV3 trains a recurrent WORLD_MODEL in supervised fashion

>     using real environment interactions (sampled from a replay buffer). The world model's objective

Reviewer: ```suggestion
    using real environment interactions sampled from a replay buffer. The world model's objective
```

    is to correctly predict the transition dynamics of the RL environment: next observation, reward,
    and a boolean continuation flag.

>     The actor- and critic-networks are subsequently trained on synthesized trajectories only,

Reviewer: ```suggestion
    RLlib trains the actor- and critic-networks on synthesized trajectories only,
```

    :width: 750

    **BC architecture:** RLlib's behavioral cloning (BC) uses Ray Data to tap into its parallel data

>     processing capabilities. In one training iteration, episodes are read in parallel from

Reviewer: ```suggestion
    processing capabilities. In one training iteration, RLlib reads episodes in parallel from
```

>     offline (ex. JSON) files by the n DataWorkers. These episodes are then preprocessed into train

Reviewer: ```suggestion
    offline, for example JSON, files by the n DataWorkers. RLlib then preprocesses these episodes into training
```

>     batches and sent as data iterators directly to the n Learners, which perform the forward- and backward passes

Reviewer: ```suggestion
    batches and sends them as data iterators directly to the n Learners, which perform the forward and backward passes
```

   :members: training

.. _cql:

Conservative Q-Learning (CQL)
-----------------------------
`[paper] <https://arxiv.org/abs/2006.04779>`__
`[implementation] <https://github.com/ray-project/ray/blob/master/rllib/algorithms/cql/cql.py>`__

.. figure:: images/algos/cql-architecture.svg
    :width: 750

    **CQL architecture:** CQL (Conservative Q-Learning) is an offline RL algorithm that mitigates the overestimation of Q-values
    outside the dataset distribution through a conservative critic estimate. It adds a simple Q regularizer loss to the standard

>     Bellman update loss, ensuring that the critic doesn't output overly optimistic Q-values. This conservative

Reviewer: ```suggestion
    Bellman update loss, ensuring that the critic doesn't output overly optimistic Q-values. You can add this conservative
```

>     correction term can be added on top of any off-policy Q-learning algorithm (here, we provide this for SAC).

Reviewer: ```suggestion
    correction term to any off-policy Q-learning algorithm. The Ray team provides this for SAC.
```

**Tuned examples:**
`Pendulum-v1 <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/cql/pendulum_cql.py>`__

> **CQL-specific configs** (see also `common configs <rllib-training.html#common-parameters>`__):

Reviewer: ```suggestion
**CQL-specific configs** and `common configs <rllib-training.html#common-parameters>`__):
```

   :members: training

.. _cql:

Conservative Q-Learning (CQL)
-----------------------------
`[paper] <https://arxiv.org/abs/2006.04779>`__
`[implementation] <https://github.com/ray-project/ray/blob/master/rllib/algorithms/cql/cql.py>`__

.. figure:: images/algos/cql-architecture.svg
    :width: 750

    **CQL architecture:** CQL (Conservative Q-Learning) is an offline RL algorithm that mitigates the overestimation of Q-values
    outside the dataset distribution through a conservative critic estimate. It adds a simple Q regularizer loss to the standard

>     Bellman update loss, ensuring that the critic doesn't output overly optimistic Q-values. This conservative
>     correction term to any off-policy Q-learning algorithm. The Ray team provides this for SAC.

Author: ```suggestion
    Bellman update loss, ensuring that the critic doesn't output overly optimistic Q-values.
    The `SACLearner` adds this conservative correction term to the TD-based Q-learning loss.
```

    **MARWIL architecture:** MARWIL is a hybrid imitation learning and policy gradient algorithm suitable for training on
    batched historical data. When the ``beta`` hyperparameter is set to zero, the MARWIL objective reduces to plain

>     imitation learning (see `BC`_). MARWIL uses Ray.Data to tap into its parallel data

Reviewer: ```suggestion
    imitation learning. See `BC`_. MARWIL uses Ray.Data to tap into its parallel data
```

>     processing capabilities. In one training iteration, episodes are read in parallel from offline files

Reviewer: ```suggestion
    processing capabilities. In one training iteration, RLlib reads episodes in parallel from offline files,
```

>     (ex. `parquet <https://parquet.apache.org/>`__) by the n DataWorkers, then preprocessed into train batches and

Reviewer: ```suggestion
    for example, `parquet <https://parquet.apache.org/>`__), by the n DataWorkers, then RLlib preprocesses them into training batches and
```

>     sent as data iterators directly to the n Learners, which perform the forward/backward passes and optimizer step.

Reviewer: ```suggestion
    sends them as data iterators directly to the n Learners, which perform the forward and backward passes and the optimizer step.
```

.. figure:: images/algos/curiosity-architecture.svg
    :width: 850

    **Intrinsic Curiosity Model (ICM) architecture:** The main idea behind ICM is to train a world-model
    (in parallel to the "main" policy) to predict the environment's dynamics. The loss of

>     the world model is the intrinsic reward that's added to the env's reward. This makes sure

Reviewer: ```suggestion
    the world model is the intrinsic reward that RLlib adds to the env's reward. This behavior makes sure
```

>     the world model is the intrinsic reward that's added to the env's reward. This makes sure

Author: ```suggestion
    the world model is the intrinsic reward that the `ICMLearner` adds to the env's
    (extrinsic) reward. This makes sure
```
